### ISSUE TITLE
XONSH_DEBUG Doesn't Disable Amalgam

### ISSUE THREAD

**Issue by mitnk** [xonsh/xonsh#1413]
```
$ a = b
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/xonsh/__amalgam__.py", line 13260, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/xonsh/__amalgam__.py", line 1907, in run_compiled_code
    func(code, glb, loc)
  File "/Users/hugo/.local/share/xonsh_ext/xonshrc.xsh", line 1, in <module>
    aliases['cda'] = ['cd', '~/foo/bar']
NameError: name 'b' is not defined

$ $XONSH_DEBUG
1
```

**Comment by melund** [comment]
It must be defined before starting xonsh. It doesn't work setting it in .xonshrc for example.

**Comment by mitnk** [comment]
All ENVs for xonsh should be put in `.xonshrc` I suppose, no?

**Comment by mitnk** [comment]
[The doc] does mentioned "Setting this variable prior to stating xonsh will supress amalgamated imports", but how? where should I set it?

**Comment by laerus** [comment]
@mitnk if you start xonsh through bash you can do `XONSH_DEBUG=1 python3 -m xonsh' , not sure how you can do it if you have xonsh as default shell

**Comment by mitnk** [comment]
@laerus thanks! I tried in xonsh like this :)

```
with ${...}.swap(XONSH_DEBUG=1):
    xonsh
```

**Comment by laerus** [comment]
was about to comment about doing it inside xonsh but when i tried it i get an error:
```
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/terminal/vt100_output.py", line 317, in _get_size
    fcntl.ioctl(fileno, termios.TIOCGWINSZ, buf, True)
OSError: [Errno 25] Inappropriate ioctl for device
```
did it worked for you?

**Comment by mitnk** [comment]
Yes, my version works for me both readline shell and PTK. What's your xonfig?

**Comment by laerus** [comment]
Btw it would be cool if we could set `XONSH_DEBUG` in either .xonshrc or config.json files

**Comment by laerus** [comment]
```
+------------------+----------------+
| xonsh            | 0.4.2.dev191   |
| Git SHA          | g8f517f7       |
| Python           | 3.5.2          |
| PLY              | 3.7            |
| have readline    | True           |
| prompt toolkit   | 1.0.0          |
| shell type       | prompt_toolkit |
| pygments         | 2.1.3          |
| on posix         | True           |
| on linux         | True           |
| distro           | debian         |
| on darwin        | False          |
| on windows       | False          |
| on cygwin        | False          |
| is superuser     | False          |
| default encoding | utf-8          |
+------------------+----------------+
```
the error happens only if i use  `$()` like `$(python3 -m xonsh)`
maybe i should open a new issue

**Comment by mitnk** [comment]
Yes, I can reproduce this. Error when using PTK, stuck when using readline. But `$[]` works ok.

**Comment by scopatz** [comment]
@laerus That looks like a bug in prompt-toolkit, but to be sure, can you post the full traceback please?

> Btw it would be cool if we could set XONSH_DEBUG in either .xonshrc or config.json files

You can, but it won't be possible to disable amalgamated imports like this.  The problem is that all of xonsh's environment handling stuff is inside of the xonsh package. So by the time you are telling xonsh to import without amalgamation, you have already imported xonsh and it is cached in sys.modules.  

I am not sure it is possible to (safely) replace these imports once they have happened.

**Comment by scopatz** [comment]
I guess we could try though.

**Comment by santagada** [comment]
why not just have it as an argument xonsh --no-amalgamation or something, that way it can be set before entering.

**Comment by scopatz** [comment]
Because the xonsh main function / argument parser is amalgamated.  However, something like that could work for xon.sh, where we already have a layer of indirection.

**Comment by scopatz** [comment]
Also, this could work for xonsh.cmd too.

**Comment by laerus** [comment]
@scopatz 

```
$ $(python3 -m xonsh)
Traceback (most recent call last):
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/interface.py", line 385, in run
    self._redraw()
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/interface.py", line 344, in _redraw
    self.renderer.render(self, self.layout, is_done=self.is_done)
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/renderer.py", line 391, in render
    size = output.get_size()
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/terminal/vt100_output.py", line 352, in get_size
    rows, columns = _get_size(stdout.fileno())
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/terminal/vt100_output.py", line 317, in _get_size
    fcntl.ioctl(fileno, termios.TIOCGWINSZ, buf, True)
OSError: [Errno 25] Inappropriate ioctl for device

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/laerus/projects/xonsh/xonsh/__main__.py", line 2, in <module>
    main()
  File "/home/laerus/projects/xonsh/xonsh/__amalgam__.py", line 16713, in main
    shell.shell.cmdloop()
  File "/home/laerus/projects/xonsh/xonsh/ptk/shell.py", line 115, in cmdloop
    line = self.singleline(auto_suggest=auto_suggest)
  File "/home/laerus/projects/xonsh/xonsh/ptk/shell.py", line 84, in singleline
    line = self.prompter.prompt(**prompt_args)
  File "/home/laerus/projects/xonsh/xonsh/ptk/shortcuts.py", line 109, in prompt
    document = cli.run(reset_current_buffer=False)
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/interface.py", line 397, in run
    self._redraw()
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/interface.py", line 344, in _redraw
    self.renderer.render(self, self.layout, is_done=self.is_done)
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/renderer.py", line 391, in render
    size = output.get_size()
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/terminal/vt100_output.py", line 352, in get_size
    rows, columns = _get_size(stdout.fileno())
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/terminal/vt100_output.py", line 317, in _get_size
    fcntl.ioctl(fileno, termios.TIOCGWINSZ, buf, True)
OSError: [Errno 25] Inappropriate ioctl for device
'\x1b]2;laerus@metahive: ~/projects/xonsh | xonsh\x07\x1b[6n\n'
```

should i open a new issue about this?

**Comment by scopatz** [comment]
@laerus, you can yes.  But I think that this might be a legitimate error.  But we can look into it more deeply in another issue.

**Comment by gforsyth** [comment]
Can this issue be closed?  I think the nature of how to set `XONSH_DEBUG` can be raised separately, but it _does_ disable amalgamation if Truthy, yes?

**Comment by scopatz** [comment]
I agree @gforsyth. let's open up a different issue
